java/vector/src/main/java/org/apache/arrow/vector/dictionary/UnionSubfieldEncoder.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.dictionary;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.complex.UnionVector;
+
+/**
+ * Sub fields encoder/decoder for Dictionary encoded {@link UnionVector}.
+ * Notes that child vectors within struct vector can either be dictionary encodeable or not.
+ */
+public class UnionSubfieldEncoder extends StructSubfieldEncoder {
```

Comment:
this class hierarchy does not make sense.  UnionEncoders don't have an "is a" relationship with StructSubfieldEncoder.

Reply:
Right, maybe I could extract the same logic in somewhere else, and remove this extends relation.

Reply:
I think the to make this (and repeated fields) generic to program against FieldVector, get all the child vectors, try to encode them and then reconstruct the vector.   Wouldn't there also be an issue creating a dictionary encoded union vectors because the FieldVectors take on the integer type corresponding to the encoding?

Reply:
Sure, the encoded UnionVector would have several integer type child vectors(in it's inner structVector), this seems not proper for UnionVector but seems no other choice unless we change the original vector type(i.e. UnionVector->StructVector). We expect this encoded vector only used for decoding(could get each integer vector via getChildrenFromFields API) and not for other purposes.